File: integrations/java/otel/src/main/java/com/openakta/logger/otel/OtelSink.java
package com.openakta.logger.otel;

import com.openakta.logger.WideEventPayload;
import com.openakta.logger.sinks.Sink;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

public final class OtelSink implements Sink {
  private final OtelLoggerLike logger;

  public OtelSink(OtelLoggerProviderLike provider) {
    this.logger = provider.getLogger("openakta-logger", "0.1.0");
  }

  @Override
  public CompletableFuture<Void> export(WideEventPayload event) {
    Map<String, Object> attributes = new LinkedHashMap<>();
    attributes.put("openakta.event_id", event.eventId());
    attributes.put("openakta.operation", event.operation());
    attributes.put("openakta.status", event.status());
    attributes.put("openakta.duration_ms", event.durationMs());
    attributes.put("service.name", event.service());
    attributes.put("deployment.environment.name", event.environment());

    for (var entry : event.context().entrySet()) {
      attributes.put("openakta.ctx." + entry.getKey(), entry.getValue());
    }

    if (event.error().message() != null) {
      attributes.put("exception.type", event.error().type());
      attributes.put("exception.message", event.error().message());
      attributes.put("exception.stacktrace", event.error().stack());
    }

    Map<String, Object> record = new LinkedHashMap<>();
    record.put("severityNumber", severityNumber(event.level()));
    record.put("severityText", event.level().toUpperCase());
    record.put("body", event.operation());
    record.put("attributes", attributes);
    record.put("timestamp", java.time.Instant.parse(event.timestampEnd()).toEpochMilli() * 1_000_000);
    record.put("observedTimestamp", java.time.Instant.parse(event.timestampStart()).toEpochMilli() * 1_000_000);
    record.put(
        "resource",
        Map.of(
            "attributes",
            Map.of(
                "service.name", event.service(),
                "deployment.environment.name", event.environment())));

    logger.emit(record);
    return CompletableFuture.completedFuture(null);
  }

  private static int severityNumber(String level) {
    return switch (level) {
      case "info" -> 9;
      case "warn" -> 13;
      case "error" -> 17;
      case "fatal" -> 21;
      default -> 9;
    };
  }
}

File: integrations/python/otel/src/openakta_logger_otel/otel_sink.py
from __future__ import annotations

from typing import Any, Protocol

from openakta_logger import Sink


class LoggerLike(Protocol):
    def emit(self, record: dict[str, Any]) -> None: ...


class ProviderLike(Protocol):
    def get_logger(self, name: str, version: str | None = None) -> LoggerLike: ...


class OtelSink(Sink):
    def __init__(self, provider: ProviderLike, logger_name: str = "openakta-logger") -> None:
        self._logger = provider.get_logger(logger_name, "0.1.0")

    async def export(self, event: dict[str, Any]) -> None:
        attributes: dict[str, Any] = {
            "openakta.event_id": event["event_id"],
            "openakta.operation": event["operation"],
            "openakta.status": event["status"],
            "openakta.duration_ms": event["duration_ms"],
            "service.name": event["service"],
            "deployment.environment.name": event["environment"],
        }

        for key, value in event["context"].items():
            attributes[f"openakta.ctx.{key}"] = value

        if event["error"]["message"]:
            attributes["exception.type"] = event["error"]["type"]
            attributes["exception.message"] = event["error"]["message"]
            attributes["exception.stacktrace"] = event["error"]["stack"]

        self._logger.emit(
            {
                "severityNumber": _severity_number(event["level"]),
                "severityText": event["level"].upper(),
                "body": event["operation"],
                "attributes": attributes,
                "timestamp": _iso_to_nanos(event["timestamp_end"]),
                "observedTimestamp": _iso_to_nanos(event["timestamp_start"]),
                "resource": {
                    "attributes": {
                        "service.name": event["service"],
                        "deployment.environment.name": event["environment"],
                    }
                },
            }
        )


def _iso_to_nanos(timestamp: str) -> int:
    from datetime import datetime

    return int(datetime.fromisoformat(timestamp.replace("Z", "+00:00")).timestamp() * 1_000_000_000)


def _severity_number(level: str) -> int:
    mapping = {"info": 9, "warn": 13, "error": 17, "fatal": 21}
    return mapping.get(level, 9)
